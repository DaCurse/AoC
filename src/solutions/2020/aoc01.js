export function partOne(inputs) {
  // We will store numbers in a set until we find our solution
  const prev = new Set();
  const target = 2020;

  for (let i = 0; i < inputs.length; ++i) {
    const input = inputs[i];

    // If a number exists in the set who is the difference between the target
    // and the current number, we found our pair
    if (!prev.has(target - input)) {
      prev.add(input);
    } else {
      return {
        // Returning the pair alongisde the result for propery testing
        fluff: [target - input, input],
        result: (target - input) * input,
      };
    }
  }
}
