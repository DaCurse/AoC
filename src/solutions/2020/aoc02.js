const splitRegex = /-|\s|: /;

/**
 * @param {string[]} inputs Array of policies
 */
export function partOne(inputs) {
  return inputs.reduce((acc, input) => {
    // Parse line using regex
    const [min, max, letter, pw] = input.split(splitRegex);
    // Count occurances
    const count = pw.split(letter).length - 1;

    // Incrementing accumulator if count is within the range
    return count >= Number(min) && count <= Number(max) ? acc + 1 : acc;
  }, 0);
}

/**
 * @param {string[]} inputs Array of policies
 */
export function partTwo(inputs) {
  return inputs.reduce((acc, input) => {
    // Parse line using regex
    const [pos1, pos2, letter, pw] = input.split(splitRegex);

    // Use XOR to determine if only one of the statements is true and incremenet
    // accumulator accordingly
    return (pw[Number(pos1) - 1] === letter) ^ (pw[Number(pos2) - 1] === letter)
      ? acc + 1
      : acc;
  }, 0);
}
