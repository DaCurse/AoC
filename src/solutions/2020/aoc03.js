export function partOne(inputs) {
  const lineLength = inputs[0].length;
  let x = 0;
  let y = 0;
  let trees = 0;
  while (y < inputs.length - 1) {
    x = (x + 3) % lineLength;
    if (inputs[++y][x] === '#') {
      trees++;
    }
  }

  return trees;
}
