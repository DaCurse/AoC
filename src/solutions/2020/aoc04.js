export function parseFields(raw) {
  return raw.split('\n\n').map((p) => p.split(/\s|:/));
}

export function partOne(inputs) {
  const requiredFields = ['byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid'];
  return inputs.reduce(
    (acc, passport) =>
      requiredFields.every((field) => passport.includes(field)) ? acc + 1 : acc,
    0
  );
}
