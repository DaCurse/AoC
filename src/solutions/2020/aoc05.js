// TODO: Finish docs

const map = {
  B: '1',
  R: '1',
  F: '0',
  L: '0',
};

function calcIDs(inputs) {
  return inputs.map((input) =>
    parseInt(
      input.replace(/[BR]|[FL]/g, (char) => map[char]),
      2
    )
  );
}

export function partOne(inputs) {
  return Math.max(...calcIDs(inputs));
}

// eslint-disable-next-line no-unused-vars
export function partTwo(inputs) {
  // TODO: Re-write a clean solution
}
