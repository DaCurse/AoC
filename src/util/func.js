/**
 * Returns the sum of `n1` and `n2`
 * @param {number} n1
 * @param {number} n2
 */
export function sum(n1, n2) {
  return n1 + n2;
}

/**
 * Returns the product of `n1` and `n2`
 * @param {number} n1
 * @param {number} n2
 */
export function product(n1, n2) {
  return n1 * n2;
}

/**
 * Returns the difference between `n1` and `n2`
 * @param {number} n1
 * @param {number} n2
 */
export function diff(n1, n2) {
  return n1 - n2;
}

/**
 * Returns the quotient from dividing `n1` by `n2`
 * @param {number} n1
 * @param {number} n2
 */
export function quotient(n1, n2) {
  return n1 / n2;
}
