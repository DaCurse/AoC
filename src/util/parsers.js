/**
 * Parses a list of strings to an array with a newline as the delimiter, getting
 * rid of empty results
 * @param {string} raw List of strings
 */
export function parseListOfStrings(raw) {
  return raw.split(/\r?\n/).filter((s) => s);
}

/**
 * Parses a list of groups seperated each by a blank line
 * @param {string} raw List of groups
 */
export function parseListOfGroups(raw) {
  return raw.split(/(?:\r?\n){2}/).filter((s) => s);
}

/**
 * Parses list of strings to an array of integers
 * @param {string} raw List of strings
 */
export function parseListOfInts(raw) {
  return parseListOfStrings(raw).map((n) => parseInt(n, 10));
}
