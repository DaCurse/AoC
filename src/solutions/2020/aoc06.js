import { parseListOfStrings } from '../../util/parsers';

export function partOne(inputs) {
  return inputs.reduce(
    (total, input) => total + new Set(input.replace(/\s/g, '')).size,
    0
  );
}

export function partTwo(inputs) {
  return inputs.reduce((total, input) => {
    const unanimous = parseListOfStrings(input)
      .map((s) => Array.from(s))
      .reduce((intersection, answers) =>
        intersection.filter((a) => answers.includes(a))
      );
    return total + unanimous.length;
  }, 0);
}
