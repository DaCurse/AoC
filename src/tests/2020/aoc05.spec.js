import { partOne, partTwo } from '../../solutions/2020/aoc05';
import getInputs from '../../util/inputs';
import { parseListOfStrings } from '../../util/parsers';

describe('Day 5 solution', () => {
  test('Part one: Find the highest ID seat', async () => {
    const inputs = await getInputs(5, parseListOfStrings);
    const result = partOne(inputs);
    console.log('Day 5 Part 1 solution:', result);

    expect(result).toMatchSnapshot();
  });

  test('Part one: Find your seat (The missing seat)', async () => {
    const inputs = await getInputs(5, parseListOfStrings);
    const result = partTwo(inputs);
    console.log('Day 5 Part 2 solution:', result);

    expect(result).toMatchSnapshot();
  });
});
