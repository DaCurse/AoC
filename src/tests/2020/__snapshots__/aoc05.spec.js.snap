// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Day 5 solution Part one: Find the highest ID seat 1`] = `896`;

exports[`Day 5 solution Part one: Find your seat (The missing seat) 1`] = `659`;
